includes/validation/class-amp-validation-manager.php

```diff
@@ -1886,54 +1798,153 @@ public static function remove_illegal_source_stack_comments( Document $dom ) {
+				$kept_count++;
+			}
+			if ( ! ( (int) $sanitization['status'] & AMP_Validation_Error_Taxonomy::ACKNOWLEDGED_VALIDATION_ERROR_BIT_MASK ) ) {
+				$unreviewed_count++;
+			}
+			$total_count++;
+		}
+
 		/*
 		 * Override AMP status in admin bar set in \AMP_Validation_Manager::add_admin_bar_menu_items()
 		 * when there are validation errors which have not been explicitly accepted.
 		 */
-		if ( ! is_admin_bar_showing() || ! self::$amp_admin_bar_item_added ) {
-			return;
+		if ( is_admin_bar_showing() && self::$amp_admin_bar_item_added ) {
+			self::update_admin_bar_item( $dom, $total_count, $kept_count, $unreviewed_count );
 		}
 
-		$error_count = 0;
-		foreach ( self::$validation_results as $validation_result ) {
-			$validation_status = AMP_Validation_Error_Taxonomy::get_validation_error_sanitization( $validation_result['error'] );
-
-			$is_unaccepted = 'with_preview' === $validation_status['forced'] ?
-				AMP_Validation_Error_Taxonomy::VALIDATION_ERROR_ACK_ACCEPTED_STATUS !== $validation_status['status']
-				:
-				AMP_Validation_Error_Taxonomy::VALIDATION_ERROR_ACK_ACCEPTED_STATUS !== $validation_status['term_status'];
-			if ( $is_unaccepted ) {
-				$error_count++;
-			}
+		// If no invalid markup is kept, then the page should definitely be displayed to the user.
+		if ( 0 === $kept_count ) {
+			return true;
 		}
 
-		if ( 0 === $error_count ) {
+		// When overrides are present, go ahead and display to the user.
+		if ( ! empty( self::$validation_error_status_overrides ) ) {
+			return true;
+		}
+
+		/*
+		 * In AMP-first, strip html@amp attribute to prevent GSC from complaining about a validation error
+		 * already surfaced inside of WordPress. This is intended to not serve dirty AMP, but rather a
+		 * non-AMP document (intentionally not valid AMP) that contains the AMP runtime and AMP components.
+		 *
+		 * Otherwise, if in Paired AMP then redirect to the non-AMP version if the current user isn't an user who
+		 * can manage validation error statuses (access developer tools) and change the AMP options for the template
+		 * mode. Such users should be able to see kept invalid markup on the AMP page even though it is invalid.
+		 */
+		if ( amp_is_canonical() ) {
+			$dom->documentElement->removeAttribute( Attribute::AMP );
+			$dom->documentElement->removeAttribute( Attribute::AMP_EMOJI );
+			$dom->documentElement->removeAttribute( Attribute::AMP_EMOJI_ALT );
+
+			/*
+			 * Make sure that document.write() is disabled to prevent dynamically-added content (such as added
+			 * via amp-live-list) from wiping out the page by introducing any scripts that call this function.
+			 */
+			$script = $dom->createElement( Tag::SCRIPT );
+			$script->appendChild( $dom->createTextNode( 'document.addEventListener( "DOMContentLoaded", function() { document.write = function( text ) { throw new Error( "[AMP-WP] Prevented document.write() call with: "  + text ); }; } );' ) );
+			$dom->head->appendChild( $script );
+			return true;
```

Comment:
This sounds like it might inadvertently break needed behavior. Are there any valid uses for `document.write` at this point?

Reply:
Calling `document.write()` after the page has loaded is never valid. It causes the page to be destroyed. Example: https://jsfiddle.net/westonruter/hq98er6j/

This prevents a scenario where an `amp-live-list` is used on the page (in an AMP-first site), and a site has _kept_ an invalid script that does `document.write()`. When the new content for the `amp-live-list` is loaded, if there was a `document.write()` call it would break the page.

Reply:
It's important to note that this function is only overridden at `DOMContentLoaded`, so it will still work properly when it should be.